Use the pki tool to bootstrap certificates during installation

rcritten opened the pull request:
We previously called out to certmonger to have it directly
obtain certificates from the CA. Instead use the CA-generated
/root/ca-admin.p12 certificate to authenticate using the pki
tool to generate the IPA RA certificate.

Then use that certificate to issue the DS, HTTP and KDC
certificates.

Fixes: https://pagure.io/freeipa/issue/9738

rcritten commented:
At least some of the tests are failing in ipa-server-upgrade with an error calling the mod_nss to mod_ssl migration. I have a separate PR to drop that https://github.com/freeipa/freeipa/pull/7671 

I'll pull that patch in here to see if everything will pass.

rcritten commented:
I had the UPN for the PKINIT certificate set incorrectly. The idp test should pass now.

rcritten commented:
This seems ready to go except for one thing. I have the mod_nss changes in a separate PR, https://github.com/freeipa/freeipa/pull/7671
So either that needs to be merged or I combine its contents more fully along with this PR.

rcritten commented:
Added additional patches to add key size flexibility. Only supports RSA keys at the moment.

frasertweedale commented on ipaserver/install/installutils.py:
```diff
@@ -1590,3 +1590,55 @@ def get_replication_plugin_name(dirsrv_get_entry):
+    size = size.strip()
+
+    if type not in types:
+        return _('Must be of one of %(types)s') % {'types': ','.join(types)}
```
Suggest the following minor UX improvement:

```python
if len(types) == 1:
    return _('Key type must be "%(type)s"') % {'type': next(iter(types))}
else:
    return _('Key type must be one of: %(types)s') % {'types': ','.join(types)}
```

frasertweedale commented on ipaserver/install/httpinstance.py:
```diff
@@ -337,26 +338,26 @@ def __setup_ssl(self):
-                ]
-                helper = " ".join(ca_args)
-                prev_helper = certmonger.modify_ca_helper('IPA', helper)
+                with tempfile.TemporaryDirectory() as tmpdir:
```
ruh roh:
```
2025-02-27T20:00:22Z DEBUG   [11/21]: setting up ssl
2025-02-27T20:00:22Z DEBUG Traceback (most recent call last):
...
  File "/usr/lib/python3.13/site-packages/ipaserver/install/httpinstance.py", line 342, in __setup_ssl
    with tempfile.TemporaryDirectory() as tmpdir:
         ^^^^^^^^
NameError: name 'tempfile' is not defined. Did you mean: 'compile'? Or did you forget to import 'tempfile'?
...
  File "/usr/lib/python3.13/site-packages/ipaserver/install/httpinstance.py", line 342, in __setup_ssl
    with tempfile.TemporaryDirectory() as tmpdir:
         ^^^^^^^^
NameError: name 'tempfile' is not defined. Did you mean: 'compile'? Or did you forget to import 'tempfile'?
```

rcritten replied:
This is fixed. The "remove mod_nss migration" change that was included with this PR was merged upstream so I dropped it from this one. That removed the tempfile import. There was a brief period where this PR was missing the import. You lucked out and found it.

frasertweedale commented on ipaserver/install/httpinstance.py:
```diff
@@ -337,26 +338,26 @@ def __setup_ssl(self):
                 # In migration case, if CA server is older version it may not
                 # have codepaths to support the ipa-ca.$DOMAIN dnsName in HTTP
-                # cert.  Therefore if request fails, try again without the
+                # cert.  Therefore if request fails, try again without them
```
I think this change is possibly a mistake?

frasertweedale reviewed (changes requested):
One error found (missing import), and some more minor feedback and suggestions.

frasertweedale commented:
I think it would be good to extract the py2 changes and key type feature to separate PRs, and minimise this bootstrap PR.

rcritten commented on ipaserver/install/installutils.py:
```diff
@@ -1590,3 +1590,55 @@ def get_replication_plugin_name(dirsrv_get_entry):
+    try:
+        ret = api.Command['config_show']()
+    except errors.PublicError as e:
+        logger.error('Cannot retrieve key type/size %s', e)
```
It's not that it isn't set, its that the retrieval of it failed. It is handled gracefully. A message is logged and installation will try to continue.

rcritten commented:
I realized I hadn't applied a couple of your suggestions. Done now.

I also recognize that with my key size/type proposal the HTTP, DS, PKINIT and RA keys will be homogeneous. All the same type and size. I'm not aware of any benefit of allowing uniqueness for each. I suspect that doing that would require quite a bit more code and increase the likelihood of user errors configuring it.

frasertweedale commented:
> I also recognize that with my key size/type proposal the HTTP, DS, PKINIT and RA keys will be homogeneous. All the same type and size. I'm not aware of any benefit of allowing uniqueness for each. I suspect that doing that would require quite a bit more code and increase the likelihood of user errors configuring it.

I thought about this too.  Some day some customer might (imagine they) want it, but it's not worth the effort at this stage.

frasertweedale commented:
> I think it would be good to extract the py2 changes and key type feature to separate PRs, and minimise this bootstrap PR.

The py2 cleanup PR: https://github.com/freeipa/freeipa/pull/7719 .

frasertweedale commented:
I've reviewed and tested the key size feature - LGTM.  And based on my testing, the certificate issuance during installation is also working fine.  What is left (for me to do) is convert to using the pki Python library rather than `pki` CLI program.

rcritten commented:
Rebased and squashed patches by myself and Fraser. The IPA installer will now use the python API rather than certmonger to obtain the initial server certificates. This lets us avoid calling the REST and XML API directly via certmonger and lets PKI determine the best method to obtain them.

Bumped PKI requires to 11.6.0. This should be sufficient for installation purposes. 11.7.0 will be needed in the future.
